From 8029903397036f8d7de34e15ede5800194e47d3b Mon Sep 17 00:00:00 2001
From: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Date: Wed, 13 Mar 2019 19:12:47 +0200
Subject: [PATCH 477/741] staging: fsl-dpaa2/mac: add pause frames support for
 managed phys

commit 1f10c93e50237f7ffaee34b2e9c32b6ee69fddb9 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Read the configured dpmac options and depending on the supported
features set the according advertising bit.

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
Signed-off-by: Bogdan Purcareata <bogdan.purcareata@nxp.com>
Signed-off-by: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>

squash! staging: fsl-dpaa2/mac: add pause frames support for managed phys
---
 drivers/staging/fsl-dpaa2/mac/mac.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/staging/fsl-dpaa2/mac/mac.c b/drivers/staging/fsl-dpaa2/mac/mac.c
index bc7bf8cc72e1..577bfb731b0a 100644
--- a/drivers/staging/fsl-dpaa2/mac/mac.c
+++ b/drivers/staging/fsl-dpaa2/mac/mac.c
@@ -149,6 +149,13 @@ static void dpaa2_mac_link_changed(struct net_device *netdev)
 		if (phydev->autoneg)
 			state.options |= DPMAC_LINK_OPT_AUTONEG;
 
+		if (phydev->pause &&
+			linkmode_test_bit(ETHTOOL_LINK_MODE_Pause_BIT, phydev->advertising))
+			state.options |= DPMAC_LINK_OPT_PAUSE;
+		if (phydev->pause &&
+			linkmode_test_bit(ETHTOOL_LINK_MODE_Asym_Pause_BIT, phydev->advertising))
+			state.options |= DPMAC_LINK_OPT_ASYM_PAUSE;
+
 		netif_carrier_on(netdev);
 	} else {
 		netif_carrier_off(netdev);
@@ -411,6 +418,20 @@ static void configure_link(struct dpaa2_mac_priv *priv,
 	phydev->speed = cfg->rate;
 	phydev->duplex  = !!(cfg->options & DPMAC_LINK_OPT_HALF_DUPLEX);
 
+	if (linkmode_test_bit(ETHTOOL_LINK_MODE_Pause_BIT, phydev->supported)) {
+		if (cfg->options & DPMAC_LINK_OPT_PAUSE)
+			linkmode_set_bit(ETHTOOL_LINK_MODE_Pause_BIT, phydev->advertising);
+		else
+			linkmode_clear_bit(ETHTOOL_LINK_MODE_Pause_BIT, phydev->advertising);
+	}
+
+	if (linkmode_test_bit(ETHTOOL_LINK_MODE_Asym_Pause_BIT, phydev->supported)) {
+		if (cfg->options & DPMAC_LINK_OPT_ASYM_PAUSE)
+			linkmode_set_bit(ETHTOOL_LINK_MODE_Asym_Pause_BIT, phydev->advertising);
+		else
+			linkmode_clear_bit(ETHTOOL_LINK_MODE_Asym_Pause_BIT, phydev->advertising);
+	}
+
 	if (cfg->advertising != 0) {
 		linkmode_zero(phydev->advertising);
 		link_mode_dpmac2phydev(cfg->advertising, phydev->advertising);
-- 
2.17.1

